feat(docs): partition split (#1660)
questdb/questdb.io: docs/operations/capacity-planning.md

:::

> #### Auto-partitioning

Author: Amended to partition split.

From QuestDB 7.2, heavily out-of-order commits are split into two partitions to

> reduce write amplification. For a merge partition to be inserted, a partition

Reviewer: ```suggestion
reduce write amplification. When data is merged into an existing partition because of out-of-order insert the partition
```

will be split into two partitions: the prefix partition and the suffix

> partition. The `TableReader` only reads the prefix partition.

Reviewer: ```suggestion
partition.
```

> A partition split happens when both of the following are true:

Reviewer: I think it would be good to introduce a sample scenario here where daily partition here is written in the middle and that falls into writing prefix, merge, and suffix.

- The prefix partition size is bigger than the combination of the suffix
  partition and the rows to be merged
- The estimated prefix partition size on disk is higher than
  `cairo.o3.partition.split.min.size` (50MB by default).

A partition can be split into more than two sub-partitions. The last partition
piece is squashed back into the main partition when the total number of
partition pieces exceeds `cairo.o3.last.partition.max.splits` (20 by default).
For all the partition pieces except the last piece, the QuestDB engine squashes
them aggressively to maintain only 1 physical partition.

> 

Reviewer: Would be good to mention that partition parts can be seen in SHOW PARTITIONS

:::

#### Partition split

From QuestDB 7.2, heavily out-of-order commits can split the partitions into
parts to reduce write amplification. When data is merged into an existing
partition as a result of an out-of-order insert, the partition will be split
into two partitions: the prefix partition and the suffix partition.

A partition split happens when both of the following are true:

- The prefix partition size is bigger than the combination of the suffix
  partition and the rows to be merged.
- The estimated prefix partition size on disk is higher than
  `cairo.o3.partition.split.min.size` (1TB by default).

For example, the following settings and partition details can trigger partition
split:

- `cairo.o3.partition.split.min.size` set to 50MB
- `cairo.o3.last.partition.max.splits` set to 20

> - A partition `2023-01-01.1` with 1,000 rows every hour, and therefore 24,000

Reviewer: This should be a new paragraph, not a continuation of the settings. Also it would be better to introduce it with something like 

Considering an example of ...

  rows in total.
- Inserting one row with the timestamp `2023-01-01T23:00`

When the out-of-order row `2023-01-01T23:00` is inserted, the partition is split
into 2 parts:

- Prefix: `2023-01-01.1` with 23,000 rows
- Suffix (including the merged row):`2023-01-01T75959-999999.2` with 1,001 rows

> A partition can be split into more than two parts. The last part is squashed

Reviewer: ```suggestion
A partition can be split into more than two parts. For the last (year, ..., hour) partition the parts are squashed
```

back into the main partition when the total number of partition parts exceeds
`cairo.o3.last.partition.max.splits` (20 by default). For all the partition

> parts except the last part, the QuestDB engine squashes them aggressively to

Reviewer: ```suggestion 
s except the last (year, ..., hour), the QuestDB engine squashes them aggressively to
```

:::

#### Partition split

From QuestDB 7.2, heavily out-of-order commits can split the partitions into
parts to reduce write amplification. When data is merged into an existing
partition as a result of an out-of-order insert, the partition will be split
into two partitions: the prefix partition and the suffix partition.

A partition split happens when both of the following are true:

> - The prefix partition size is bigger than the combination of the suffix

Author: I think this can be simplified to just "prefix". 
```suggestion
- The prefix size is bigger than the combination of the suffix
```

  partition and the rows to be merged.

> - The estimated prefix partition size on disk is higher than

Author: Removed to just have prefix.

>   `cairo.o3.partition.split.min.size` (1TB by default).

Author: > 1TB is used to effectively disable splitting. We are going to enable it and set default to 50MB

Are we going to do it next week? We can change this from 1TB to 50 MB later, when we actually do it?

> A partition can be split into more than two parts. The last (year, ..., hour)

Author: Amended the wording a bit.

> partition is squashed back into the main partition when the total number of

Reviewer: ```suggestion
partition parts are squashed together when
```

Author: Also tidied up the sentence a bit.

partition parts exceeds `cairo.o3.last.partition.max.splits` (20 by default).
For all the partitions except the last (year, ..., hour), the QuestDB engine

> squashes them aggressively to maintain only one physical partition.

Reviewer: ```suggestion
squashes them aggressively to maintain only one physical partition at the end of every commit.
```

The SQL keyword [SHOW PARTITIONS](/docs/reference/sql/show/) can be used to

> display partition split details.

Reviewer: I would add here something like

All partition operations (DROP, ATTACH, DETACH) do not consider partition splits as individual partitions and work on the table partitioning unit (year, week, ..., hour). For example, when a daily partition consisting of several parts is dropped all the parts belonging to the given date are dropped. Similarly, when the multipart daily partition is detached it is squashed into a single piece first and then detached.

##### Partition split example

Considering an example of the following settings and partition details:

> - `cairo.o3.partition.split.min.size` set to 50MB

Reviewer: Please remove config values from the example they are not relevant
```suggestion
```
